### PR TITLE
[Error Handling/Codes][State]Starts ETAG MISMATCH Error Handling

### PR DESCRIPTION
**Pull request by robertojrojas**
# Description

Starts State ETAG MISMATCH Error Handling. 
Redis is the first component to implement the Error codes. 

**_NOTE:_** This PR depends on KIT PR: https://github.com/dapr/kit/pull/57

## Issue reference
PR is part of the Dapr Issue: [ETAG MISMATCH](https://github.com/dapr/dapr/issues/6472)

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: state/redis/redis.go [hunk 1]
```diff
@@ -117,6 +120,10 @@ func newStateStore(log logger.Logger) *StateStore {
 		json:                           jsoniter.ConfigFastest,
 		logger:                         log,
 		suppressActorStateStoreWarning: atomic.Bool{},
+		resourceInfoData: errorcodes.ResourceInfoData{
+			ResourceType: "state.redis/v1",
+			ResourceName: "REDIS",
```
Comment (ItalyPaleAle) [inline state/redis/redis.go:125]: Does this have to be ALL CAPS?

File: internal/errorcodes/errorcodes.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+var StateETagMismatchReason = Reason("DAPR_STATE_ETAG_MISMATCH")
+
+type ResourceInfoData struct {
+	ResourceType, ResourceName string
+}
+
+func FeatureEnabled(md map[string]string) bool {
+	if _, ok := md[ErrorCodesFeatureMetadataKey]; ok {
```
Comment (ItalyPaleAle) [inline internal/errorcodes/errorcodes.go:40]: As per comment on other PR, our metadata parsing should ideally be case-insensitive.

File: internal/errorcodes/errorcodes.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		return true
+	}
+	return false
+}
+
+// NewStatusError returns a Status representing Code, error Reason, Message, and optional ResourceInfo and Metadata.
+// When successful, it returns a StatusError, otherwise returns the original error
+func NewStatusError(code codes.Code, err error, errDescription string, reason Reason, rid *ResourceInfoData, metadata map[string]string) error {
```
Comment (ItalyPaleAle) [inline internal/errorcodes/errorcodes.go:48]: Why do we need both `err error` and `errDescription string`? Can't the two be inferred from one another?

File: internal/errorcodes/errorcodes.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+func FeatureEnabled(md map[string]string) bool {
+	if _, ok := md[ErrorCodesFeatureMetadataKey]; ok {
+		return true
+	}
+	return false
+}
+
+// NewStatusError returns a Status representing Code, error Reason, Message, and optional ResourceInfo and Metadata.
+// When successful, it returns a StatusError, otherwise returns the original error
```
Comment (ItalyPaleAle) [inline internal/errorcodes/errorcodes.go:47]: Please pass `rid` not as pointer, so it's a bit easier on the GC.

File: internal/errorcodes/errorcodes.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+func NewStatusError(code codes.Code, err error, errDescription string, reason Reason, rid *ResourceInfoData, metadata map[string]string) error {
+	md := metadata
```
Comment (ItalyPaleAle) [inline internal/errorcodes/errorcodes.go:49]: Why this?

File: internal/errorcodes/errorcodes.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	if md == nil {
+		md = map[string]string{}
+	}
+
+	messages := []protoiface.MessageV1{
```
Comment (ItalyPaleAle) [inline internal/errorcodes/errorcodes.go:54]: Isn't `messagev1` the interface for old protobuf messages? Which are deprecated? Can we use the v2 please?

**Review by ItalyPaleAle — COMMENTED** [review]
See comments

File: internal/errorcodes/errorcodes.go (hunk 3 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go:48]: The idea behind passing both separate is to give the calling code flexibility on the `errDescription`.

File: internal/errorcodes/errorcodes.go (hunk 5 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go:49]: I was hoping to avoid `panic: assignment to entry in nil map` down the road.

Comment (ItalyPaleAle) [reply internal/errorcodes/errorcodes.go:49]: But you don't need to copy `md := metadata`, just modify metadata :)

File: internal/errorcodes/errorcodes.go (hunk 6 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go:54]: Hmm... I used what `Status.WithDetails` takes which in my editor shows as `protoiface.MessageV1`

File: internal/errorcodes/errorcodes.go (hunk 3 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go:48]: I like your suggestion better. will change it.

File: internal/errorcodes/errorcodes.go (hunk 4 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go:47]: I wanted `rid` to be a pointer so it's _optional_

File: internal/errorcodes/errorcodes.go (hunk 3 above)
Comment (ItalyPaleAle) [reply internal/errorcodes/errorcodes.go:48]: I see that in the code you've written so far, in most cases the error is similar or derived.

how about creating a custom error object that has a "Description()" property that is optional?

```go
type ComponentError struct {
    err error
	code codes.Code
	description string
	resourceInfoData *ResourceInfoData
	metadata map[string]string
}

func NewComponentError(code codes.Code, err error) ComponentError {
	return ComponentError{
		code: code,
		err: err,
	}
}

// Error implements the error interface.
func (c ComponentError) Error() string {
	return c.err.Error()
}

// Unwrap implements the error unwrapping interface.
func (c ComponentError) Unwrap() error {
	return c.err
}

// Description returns the description of the error.
func (c ComponentError) Description() string {
	if c.description {
		return c.description
	}
	return c.err.Error()
}

func (c *ComponentError) SetDescription(description string) {
	c.description = description
}

// Add methods to access code, resourceInfoData, metadata
// Add methods to set resourceInfoData and metadata
```

Then you can pass a ComponetnError around as an `error`. In the code that handles errors, you can check:

```go
var compErr ComponentError
if errors.As(err, &compErr) {
// ...
}
```

Using custom errors is a more idiomatic thing in Go. It also allows accessing the `err` property for unwrapping.

Some links: https://www.digitalocean.com/community/tutorials/creating-custom-errors-in-go

File: internal/errorcodes/errorcodes.go (hunk 4 above)
Comment (ItalyPaleAle) [reply internal/errorcodes/errorcodes.go:47]: Ok

File: internal/errorcodes/errorcodes.go (hunk 2 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go:40]: As I mentioned in the other PR, will be shared one both PRs are merged.

File: internal/errorcodes/errorcodes.go (hunk 3 above)
Comment (robertojrojas) [reply internal/errorcodes/errorcodes.go]: Yep, custom errors are idiomatic in Go, but the idea is to follow [The proposal and use gRPC error handling: 0009-BCIRS-error-handling-codes ](https://github.com/dapr/proposals/blob/main/0009-BCIRS-error-handling-codes.md)

Comment (ItalyPaleAle) [reply internal/errorcodes/errorcodes.go]: This is still compliant with the proposal :)

You could also add the GRPCStatus method to make the custom error be a gRPC error too: https://github.com/dapr/dapr/blob/master/pkg/messages/api_error.go#L79-L83

File: internal/errorcodes/errorcodes.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+	ste, stErr := status.New(code, errDescription).WithDetails(messages...)
+	if stErr != nil {
+		return err
+	}
+
+	return ste.Err()
+}
+
+func NewErrorInfo(reason Reason, md map[string]string) *errdetails.ErrorInfo {
```
Comment (artursouza) [inline internal/errorcodes/errorcodes.go:65]: Make this private

File: internal/errorcodes/errorcodes.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	ei := errdetails.ErrorInfo{
+		Domain:   Domain,
+		Reason:   string(reason),
+		Metadata: md,
+	}
+
+	return &ei
+}
+
+func NewResourceInfo(rid *ResourceInfoData, description string) *errdetails.ResourceInfo {
```
Comment (artursouza) [inline internal/errorcodes/errorcodes.go]: make this private

File: internal/errorcodes/errorcodes.go (hunk 3 above)
Comment (artursouza) [reply internal/errorcodes/errorcodes.go]: I like the idea of having an internal struct to hold all the error information. On the other hand, I am not a fan of having it as a component specific thing. I suggest we have an internal error struct defined in kit and that being the common object for all dapr internal code. The runtime then will have a logic to transform that object into the gRPC error we agreed in the proposal.

Comment (ItalyPaleAle) [reply internal/errorcodes/errorcodes.go]: +1 on getting this in Kit. I like that

**Comment by robertojrojas** [comment]
@ItalyPaleAle @artursouza I like your proposed implementation related to Customer Errors and hosting them in the `kit`. I've created the following issue https://github.com/dapr/kit/issues/56 
I'm working on a PR now.

**Pull request closed without merging**